streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1477,6 +1590,67 @@ public Object getStateLock() {
         return stateLock;
     }
 
+    public Map<String, KafkaFuture<Uuid>> consumerClientInstanceIds(final Duration timeout) {
+        boolean setDeadline = false;
+
+        final Map<String, KafkaFuture<Uuid>> result = new HashMap<>();
+
+        KafkaFutureImpl<Uuid> future = new KafkaFutureImpl<>();
+        if (mainConsumerClientInstanceId != null) {
+            future.complete(mainConsumerClientInstanceId);
+        } else {
+            mainConsumerInstanceIdFuture = future;
+            setDeadline = true;
+        }
+        result.put(getName() + "-consumer", future);
+
+        future = new KafkaFutureImpl<>();
+        if (restoreConsumerClientInstanceId != null) {
+            future.complete(restoreConsumerClientInstanceId);
+        } else {
+            restoreConsumerInstanceIdFuture = future;
+            setDeadline = true;
+        }
+        result.put(getName() + "-restore-consumer", future);
+
+        if (setDeadline) {
+            fetchDeadline = time.milliseconds() + timeout.toMillis();
+        }
+
+        return result;
+    }
+
+    public KafkaFuture<Map<String, KafkaFuture<Uuid>>> producersClientInstanceIds(final Duration timeout) {
+        final KafkaFutureImpl<Map<String, KafkaFuture<Uuid>>> result = new KafkaFutureImpl<>();
```

Comment:
`result` future will complete only after the `Map` it returns is fully populated. For EOSv1, we cannot do this sync, thus we need this indirection.

For ALOS and EOSv2, the map is create sync (it's just a singleton for the thread producer).

Cf. other comment below.